tensorflow/compiler/xla/service/gpu/ir_emitter.cc

```diff
@@ -509,6 +515,50 @@ Status IrEmitter::EmitAtomicOperationForNestedComputation(
                                      source_address);
 }
 
+bool IrEmitter::IsEmittingForAMDGPU() const {
+  llvm::Triple target_triple = llvm::Triple(module_->getTargetTriple());
+  return target_triple.isAMDGPU();
+}
+
+bool IrEmitter::EmitAMDGPUAtomics(
+    llvm::Value* output_address, llvm::Value* source,
+    PrimitiveType element_type) {
+  if (IsEmittingForAMDGPU()) {
+    auto output_address_type =
+        llvm::dyn_cast<llvm::PointerType>(output_address->getType());
+    CHECK_NE(output_address_type, nullptr);
+
+    bool atomic_add_supported = (element_type == F32);
+    if (atomic_add_supported) {
+      auto output_ptr = [&]() {
+        if (output_address_type->getPointerAddressSpace() != 3) {
+          // the compiler will only generate a global_atomic_fadd if the pointer
+          // is in global addrspace (1)
+          return b_.CreateAddrSpaceCast(
+              output_address,
+              llvm::PointerType::get(output_address_type->getPointerElementType(),
+                                     /*AddressSpace=*/1));
+        } else {
+          // adds to shared memory are always atomic.
+          return output_address;
+        }
+      }();
+      AtomicRMW(llvm::AtomicRMWInst::FAdd, output_ptr, source,
```

Comment:
What is the main difference between NVidia and AMD for atomic calls? Is it just the casting? Maybe we could share at a lower granularity?

Reply:
it's not just the casting.  it is also the sync scope.  i would imagine that nvidia would simply use the default 'system' sync scope but we need to use 'agent' for performance reasons.

Reply:
i think the 'agent' sync scope is AMD specific.